File: solitaire/src/main/java/CardStack.java
 import java.util.Stack;

public class CardStack extends Stack<Card> {
	protected int arrayIndex;

	public void setCoords() {
	}

	public Card push(Card c) {
		c.setContainer(this);
		super.push(c);
		setCoords();
		return c;
	}

	public void setIndex(int i) {
		arrayIndex = i;
	}

	public boolean validPlacement(Card base, Card toPlace) {
		return true;
	}

	public void draw(java.awt.Graphics g) {
		if (size() != 0)
			get(size() - 1).drawAndPopulate(g);
	}
	
	public Card peek() {
		if (empty())
			return null;
		return super.peek();
	}
	
	public void populate() {
		
	}
}

File: solitaire_js/src/CardStack.js
class CardStack extends Array {
  constructor() {
    this.indexOfStack = 0;
  }

  // override
  push(c) {
    c.setContainer(this);
    setCoords();
    return super.push(c);
  }

  setIndexOfStack(i) {
    this.indexOfStack = i;
  }

  validPlacement(base, toPlace) {
    return true;
  }

  draw(ctx, cardMap, stackMap) {
    if (this.length > 0)
      this.peek().drawAndPopulate(ctx, cardMap);
  }

  peek() {
    if (this.length > 0)
      return this[this.length - 1];
    else
      return null;
  }

  populate() { }

  append(cs) {
    for (c in cs)
      this.push(c);
  }
}